File: loop_sample.py
import os
from typing import List, Tuple

def adj_path(path: str):
    if path.endswith("/"):
        pass
    else:
        path = path + "/"
    return path

def glob_paths(target_path: str, suffix: str, *, max_depth: int) -> List[str]:

    matched_paths: List[str] = []
    loop_target_paths: List[Tuple[str, int]] = []
    target_path_adj: str = adj_path(target_path)
    depth: int = 0
    loop_target_paths.append((target_path_adj, depth))

    while loop_target_paths:
        loop_target_path, depth = loop_target_paths.pop(0)
        if depth >= max_depth:
            break
        print(f'loop_target_paths:{loop_target_paths} \n')
        if os.path.isdir(loop_target_path):
            for list_item in os.listdir(loop_target_path):
                sub_path = os.path.join(loop_target_path, list_item)
                loop_target_paths.append((sub_path, depth + 1))
        elif os.path.isfile(loop_target_path):
            if loop_target_path.endswith(suffix):
                matched_paths.append(loop_target_path)
    return matched_paths

def main():
    result = glob_paths(os.getcwd(), suffix=".json", max_depth=3)
    print(f'JSONファイルの配列:{result}')
main()

File: loop_sample.rb
require 'find'

# ターゲットパスを調整する
def adjust_path(target_path)
  if target_path.end_with?('/')
    target_path
  else
    target_path + '/'
  end
end

# 指定されたパスから、指定された最大深度以下の、指定された拡張子に一致するファイルパスを取得する
def glob_paths(target_path, suffix, max_depth)
  matched_paths = [] # 一致したファイルパスを格納するための配列
  target_path = adjust_path(target_path)
  Find.find(target_path) do |current_path|
    if File.directory?(current_path) # パスがディレクトリの場合
      if File.basename(current_path) == '.' || File.basename(current_path).start_with?('.')
        Find.prune # カレントディレクトリまたは隠しファイルの場合はスキップする
      else
        depth = current_path.count('/') - target_path.count('/')
        if depth >= max_depth # 指定された深度を超えたら処理を終了する
          Find.prune
        end
      end
    elsif File.file?(current_path) && current_path.end_with?(suffix) # パスがファイルで、指定された拡張子に一致する場合
      matched_paths << current_path # 一致したファイルパスを配列に追加する
    end
  end
  matched_paths # 一致したファイルパスの配列を返す
end

result = glob_paths('.', '.json', 3)
puts "JSONファイルの配列: #{result}"
